lib/billy/cache.rb

```diff
@@ -72,11 +72,27 @@ def key(method, orig_url, body, log_key = false)
       ignore_params = Billy.config.ignore_params.include?(format_url(orig_url, true))
       merge_cached_response_key = _merge_cached_response_key(orig_url)
       url = Addressable::URI.parse(format_url(orig_url, ignore_params))
-      key = if merge_cached_response_key
-              method + '_' + Digest::SHA1.hexdigest(scope.to_s + merge_cached_response_key)
-            else
-              method + '_' + url.host + '_' + Digest::SHA1.hexdigest(scope.to_s + url.to_s)
-            end
+
+      key_pattern = Billy.config.key_pattern.select do |v|
+        Billy::Config::KEY_PATTERN_WHITELIST.include?(v)
+      end
+
+      key_base = key_pattern.each_with_object([]) do |v, base|
+        base << case v
+                when :method then method
+                when :host then url.host
+                when :scope then scope.to_s unless scope.to_s.empty?
+                else v.to_s unless v.to_s.empty?
+                end
+      end.compact
+
+      key = key_base.join('_') + '_' +
+        if merge_cached_response_key
+          Digest::SHA1.hexdigest(scope.to_s + merge_cached_response_key)
```

Comment:
Since you're adding scope to the filename, you can remove it from both of these `hexdigest`.  It was there to force a unique sha to be a different filename.